backups: add module

nixbitcoin opened the pull request:
This PR introduces nix-bitcoin's own backups module which backups up all important files in `/var/lib` to `/var/lib/localBackups` (fetchable with `nixops scp`) or a remote target url once a day using duplicity. Backups are encrypted with an apg generated password. A user can also backup the blockchains and electrs data directory with `cfg.with-bulk-data` (off-by-default).

The idea is to create a generic `backups.nix` module, which at first has duplicity support but supports extending to a variety of backup programs like restic, git-annex and borg.

Closes https://github.com/fort-nix/nix-bitcoin/issues/167

Kixunil commented on pkgs/generate-secrets/generate-secrets.sh:
```diff
@@ -11,10 +11,12 @@ makePasswordSecret lnd-wallet-password
 [[ -e lightning-charge-env ]] || echo "API_TOKEN=$(cat lightning-charge-token)" > lightning-charge-env
 [[ -e nanopos-env          ]] || echo "CHARGE_TOKEN=$(cat lightning-charge-token)" > nanopos-env
 [[ -e spark-wallet-login   ]] || echo "login=spark-wallet:$(cat spark-wallet-password)" > spark-wallet-login
+[[ -e backup-encryption-env ]] || echo "PASSPHRASE=$(cat backup-encryption-password)" > backup-encryption-env
```
This is not safe as any user running `ps aux | grep 'echo PASSPHRASE` in the right moment (can be achieved using infinite loop) will see the passphrase. Proper way to do it:

```bash
echo -n PASSPHRASE= | cat - backup-encryption-password > backup-encryption-env
```

Of course, the other lines seem to be vulnerable too. FYI, I'm not even sure if passing the passphrase via environment is safe - I've read that it isn't but I couldn't figure out how to abuse it.

erikarvstedt replied:
`echo` is a bash builtin, so its arguments are invisible outside of the script process.

Kixunil replied:
Would it make sense to write `builtin echo` in order to make double sure nothing screws up? I'm not sure how much is bash reliable in calling builtins.

erikarvstedt replied:
`echo` has been a builtin since the first Bash release in the late 80s. It's totally safe to use. :wink:

jonasnick commented on modules/backups.nix:
```diff
@@ -0,0 +1,76 @@
+
+let
+  cfg = config.services.backups;
+  duplicity-filelist = pkgs.writeText "filelist.txt" ''
```
I think it would be good if the filelist can be extended with extra lines.

Generally, the most bare-bones but still helpful version backup module is just the list of files that are important to back up. That would  then allow users to integrate nix-bitcoin with their own backup tools without having to go through the step of having the `localBackups` directory (and an additional duplicity password). Therefore it would be nice if (1) the filelist can read from other modules and (2) the filelist can be read from the operator user's shell.

nixbitcoin replied:
> I think it would be good if the filelist can be extended with extra lines.
Resolved

nixbitcoin replied:
> Generally, the most bare-bones but still helpful version backup module is just the list of files that are important to back up. That would then allow users to integrate nix-bitcoin with their own backup tools without having to go through the step of having the localBackups directory (and an additional duplicity password). Therefore it would be nice if (1) the filelist can read from other modules and (2) the filelist can be read from the operator user's shell.

The operator user can already read the filelist from the nix-bitcoin node's nix store.

We could make the filelist an option that defaults to our filelist, similar to nodeinfo script, so it can be read from all modules. I don't think it should be too difficult to translate the duplicity filelist syntax into something other programs understand.

nixbitcoin commented:
Added an extraFiles option and added `full-if-older-than` so duplicity perform a full backup if an incremental backup is requested, but the latest full backup in the collection is older than the given time, 1 month.

nixbitcoin commented:
> Concept ACK. It's nice and simple. I wonder what the long term impact on the disk space usage is. Is there an option to prune the duplicity backups?

The user can always run `duplicity remove-all-but-n-full <count> --force target_url` to prune his backups directory. This deletes all backups sets that are older than the count:th last full backup (in other words, keep the last count full backups and associated incremental sets).

nixbitcoin commented:
Fixed nits

nixbitcoin commented:
Rebased and fixed nits
